http-server/src/main/java/io/micronaut/http/server/filter/DefaultFilterBodyParser.java

```diff
@@ -0,0 +1,107 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+/**
+ * Implementation of {@link FilterBodyParser} which leverages {@link ServerHttpRequest#byteBody()} API.
+ * For form-url-encoded payloads it uses the {@link FormUrlEncodedDecoder} API.
+ * For JSON payloads it uses a {@link JsonMapper}.
+ * @author Sergio del Amo
+ * @since 4.11.0
+ */
+@Requires(beans = {JsonMapper.class, FormUrlEncodedDecoder.class})
+@Singleton
+@Experimental
+final class DefaultFilterBodyParser implements FilterBodyParser {
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultFilterBodyParser.class);
+    private final JsonMapper jsonMapper;
+    private final FormUrlEncodedDecoder formUrlEncodedDecoder;
+
+    /**
+     * @param jsonMapper JSON Mapper
+     * @param formUrlEncodedDecoder Decoder for form-url-encoded payload
+     */
+    DefaultFilterBodyParser(FormUrlEncodedDecoder formUrlEncodedDecoder,
+                            JsonMapper jsonMapper) {
+        this.formUrlEncodedDecoder = formUrlEncodedDecoder;
+        this.jsonMapper = jsonMapper;
+    }
+
+    @Override
+    @NonNull
+    @SingleResult
+    public CompletableFuture<Map<String, Object>> parseBody(@NonNull HttpRequest<?> request) {
+        Optional<MediaType> mediaTypeOptional = request.getContentType();
+        if (mediaTypeOptional.isEmpty()) {
+            LOG.debug("Could not parse body into a Map because the request does not have a Content-Type");
+            return CompletableFuture.completedFuture(Collections.emptyMap());
+        }
+        if (!(request instanceof ServerHttpRequest<?>)) {
+            LOG.debug("Could not parse body into a Map because the request is not an instance of ServerHttpRequest");
+            return CompletableFuture.completedFuture(Collections.emptyMap());
+        }
+        MediaType contentType = mediaTypeOptional.get();
+        if (contentType.equals(MediaType.APPLICATION_FORM_URLENCODED_TYPE)) {
+            return parseFormUrlEncoded((ServerHttpRequest<?>) request);
+        } else if (contentType.equals(MediaType.APPLICATION_JSON_TYPE)) {
+            return parseJson((ServerHttpRequest<?>) request);
+        }
+        LOG.debug("Could not parse body into a Map because the request's content type is not either application/x-www-form-urlencoded or application/json");
+        return CompletableFuture.completedFuture(Collections.emptyMap());
+    }
+
+    private CompletableFuture<Map<String, Object>> parseJson(@NonNull ServerHttpRequest<?> request) {
+        try (CloseableByteBody closeableByteBody = request.byteBody().split(ByteBody.SplitBackpressureMode.FASTEST)) {
+            return closeableByteBody.buffer()
+                .thenApply(bb -> {
+                    try {
+                        return jsonMapper.readValue(bb.toByteArray(), Argument.mapOf(String.class, Object.class));
```

Comment:
extract the `mapOf` call into a constant

Reply:
resolved via https://github.com/micronaut-projects/micronaut-core/pull/11272/commits/235c74d51945f9ee4b9da864cfb5039adb1027cc